### PR TITLE
 RavenDB-12161 Extension point to override OS metrics

### PR DESCRIPTION
**Pull request by haludi**
http://issues.hibernatingrhinos.com/issue/RavenDB-12161

### CODE REVIEW

File: src/Raven.Server/Config/Categories/MonitoringConfiguration.cs [hunk 1]
```diff
@@ -1,12 +1,23 @@
...
 namespace Raven.Server.Config.Categories
 {
     public class MonitoringConfiguration : ConfigurationCategory
     {
+        [Description("A command or executable to run which will provide machine cpu usage and total machine cores to standard output. If specified, RavenDB will use this information for monitoring cpu usage.")]
+        [DefaultValue(null)]
+        [ConfigurationEntry("Monitoring.Cpu.Exec", ConfigurationEntryScope.ServerWideOnly)]
+        public string CpuUsageMonitorExec { get; set; }
+
+        [Description("The command line arguments for the 'Security.MasterKey.Exec' command or executable.")]
```
Comment (iftahbe) [inline src/Raven.Server/Config/Categories/MonitoringConfiguration.cs]: forgot to change 'Security.MasterKey.Exec'

File: src/Raven.Server/Utils/CpuUsageExtensionPoint.cs [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+                Dispose();
+            }
+
+            var _ = ReadProcess(); // explicitly starting async task without waiting for it
+            _ = ReadErrors();
+        }
+
+        private async Task ReadErrors()
+        {
+            var errors = await _process.StandardError.ReadLineAsync();
```
Comment (iftahbe) [inline src/Raven.Server/Utils/CpuUsageExtensionPoint.cs]: This will read only one line of the error.

File: src/Raven.Server/Utils/CpuUsage.cs [hunk 3]
```diff
@@ -77,7 +94,7 @@ private static double CalculateProcessCpuUsage(ProcessInfo currentInfo, ProcessI
                 return _lastCpuInfo?.ProcessCpuUsage ?? 0;
             }
 
-            if (currentInfo.ActiveCores == 0)
+            if (Math.Abs(currentInfo.ActiveCores) < 0.01)
```
Comment (grisha-kotler) [inline src/Raven.Server/Utils/CpuUsage.cs]: You need to skip those checks when using the CpuUsageExtensionPoint

File: src/Raven.Server/Utils/CpuUsage.cs [hunk 4]
```diff
@@ -87,7 +104,7 @@ private static double CalculateProcessCpuUsage(ProcessInfo currentInfo, ProcessI
             }
 
             var processCpuUsage = (processorTimeDiff * 100.0) / timeDiff / currentInfo.ActiveCores;
-            if (currentInfo.ActiveCores == ProcessorInfo.ProcessorCount)
+            if ((int)currentInfo.ActiveCores == ProcessorInfo.ProcessorCount)
```
Comment (grisha-kotler) [inline src/Raven.Server/Utils/CpuUsage.cs]: same here

File: src/Raven.Server/Utils/CpuUsage.cs [hunk 5]
```diff
@@ -18,11 +18,24 @@ public static class CpuUsage
...
         private static readonly (double MachineCpuUsage, double ProcessCpuUsage) EmptyCpuUsage = (0, 0);
+        private static CpuUsageExtensionPoint _cpuUsageExtensionPoint;
+
+
+        public static CpuUsageExtensionPoint CpuUsageExtensionPoint
+        {
+            get => _cpuUsageExtensionPoint;
+            set
+            {
+                _previousExtensionPointInfo = new ExtensionPointInfo(value.Data);
```
Comment (maximburyak) [inline src/Raven.Server/Utils/CpuUsage.cs]: What do we do with the previous value, if we have one, and if we don't, then it should be just a Lazy value?

File: src/Raven.Server/Utils/CpuUsage.cs [hunk 6]
```diff
@@ -96,6 +113,24 @@ private static double CalculateProcessCpuUsage(ProcessInfo currentInfo, ProcessI
...
+            var extensionPoint = new ExtensionPointInfo(CpuUsageExtensionPoint.Data);
+            var machineCpuUsage = extensionPoint.MachineCpuUsage;
+            if (machineCpuUsage < 0)
+            {
+                if (Logger.IsOperationsEnabled)
+                {
+                    Logger.Operations($"Received bad data for cpu usage from extension point : {machineCpuUsage}");
+                }
+
+                machineCpuUsage = 0;
```
Comment (maximburyak) [inline src/Raven.Server/Utils/CpuUsage.cs]: Not sure that in this case we should use 0, maybe we should fall back to the "normal" calculation

File: src/Raven.Server/Utils/CpuUsageExtensionPoint.cs [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+                _process.Start();
+            }
+            catch (Exception e)
+            {
+                NotifyWarning("Could not start cpu usage extension point process", e);
+
+                Dispose();
+            }
+
+            var _ = ReadProcess(); // explicitly starting async task without waiting for it
```
Comment (maximburyak) [inline src/Raven.Server/Utils/CpuUsageExtensionPoint.cs]: What if we failed starting the process?

File: src/Raven.Server/Utils/CpuUsageExtensionPoint.cs [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+                _process = null;
+            }
+        }
+
+
+        private async Task ReadProcess()
+        {
+            while (IsDisposed == false)
+            {
+                var nextLine = _process.StandardOutput.ReadLineAsync();
```
Comment (maximburyak) [inline src/Raven.Server/Utils/CpuUsageExtensionPoint.cs]: We should Try-Catch here, we don't want an unhandled failed task, we could for an instance have a temporary out of memory problem here, that will just kill the task, without us knowing/monitoring it.

File: src/Raven.Server/RavenServer.cs [hunk 9]
```diff
@@ -1628,6 +1644,7 @@ public void Dispose()
                 ea.Execute(() => _clusterMaintenanceWorker?.Dispose());
 
                 ea.ThrowIfNeeded();
+                CpuUsage.CpuUsageExtensionPoint?.Dispose();
```
Comment (ayende) [inline src/Raven.Server/RavenServer.cs]: This should be inside the `ea.Execute`

File: src/Raven.Server/Dashboard/ThreadsInfo.cs [hunk 10]
```diff
@@ -14,7 +14,7 @@ public class ThreadsInfo : AbstractDashboardNotification
 
         public double CpuUsage { get; set; }
 
-        public long ActiveCores { get; set; }
+        public double ActiveCores { get; set; }
```
Comment (ayende) [inline src/Raven.Server/Dashboard/ThreadsInfo.cs]: There is no meaning from our perspective to `0.35` of a core. The extension script needs to do the transaction, not us

File: src/Raven.Server/Utils/CpuUsage.cs [hunk 11]
```diff
@@ -96,6 +113,24 @@ private static double CalculateProcessCpuUsage(ProcessInfo currentInfo, ProcessI
             return Math.Min(100, processCpuUsage);
         }
 
+        private static (double MachineCpuUsage, double ProcessCpuUsage) CalculateCpuFromExtensionPoint()
+        {
+            var extensionPoint = new ExtensionPointInfo(CpuUsageExtensionPoint.Data);
+            var machineCpuUsage = extensionPoint.MachineCpuUsage;
+            if (machineCpuUsage < 0)
```
Comment (ayende) [inline src/Raven.Server/Utils/CpuUsage.cs]: In what case this can happen? And why shouldn't we abort the extension at this point?

File: src/Raven.Server/Utils/CpuUsage.cs [hunk 12]
```diff
@@ -276,7 +311,25 @@ private static unsafe MacInfo GetMacInfo()
...
+                return GetNumberOfActiveCoresInternal(process);
+            }
+            if (extensionPointRawData.ActiveCores >= 0)
+            {
+                return extensionPointRawData.ActiveCores;
+            }
+
+            if (Logger.IsOperationsEnabled)
+            {
+                Logger.Operations($"Received bad data for active cores from extension point : {extensionPointRawData.ActiveCores}");
```
Comment (ayende) [inline src/Raven.Server/Utils/CpuUsage.cs]: This like the admin, what is `from extension point` which one, etc?
Need to be very explicit about this.

File: src/Raven.Server/Utils/CpuUsageExtensionPoint.cs (hunk 2 above)
Comment (ayende) [reply src/Raven.Server/Utils/CpuUsageExtensionPoint.cs]: Also need error handling here

File: src/Raven.Server/Utils/CpuUsageExtensionPoint.cs [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+            if (data == null)
+            {
+                HandleError("The output stream of the process has closed.");
+                return false;
+            }
+            using (_contextPool.AllocateOperationContext(out var context))
+            {
+                try
+                {
+                    var blittable = context.ReadForMemory(data, string.Empty);
```
Comment (ayende) [inline src/Raven.Server/Utils/CpuUsageExtensionPoint.cs]: Please use a descriptive name for the tag, not `string.Empty`

File: src/Raven.Server/Utils/CpuUsageExtensionPoint.cs [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+            catch (Exception)
+            {
+                // When the process terminating, killed or exited
+            }
+            finally
+            {
+                if (_didRestart == false)
+                {
+                    _didRestart = true;
+                    NotifyWarning($"{msg} \nTherefore the process will restart.", e);
```
Comment (ayende) [inline src/Raven.Server/Utils/CpuUsageExtensionPoint.cs]: use `Environment.NewLine`

File: src/Raven.Server/Utils/CpuUsageExtensionPoint.cs [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+            }
+            catch (Exception)
+            {
+                // When the process terminating, killed or exited
+            }
+            finally
+            {
+                if (_didRestart == false)
+                {
+                    _didRestart = true;
```
Comment (ayende) [inline src/Raven.Server/Utils/CpuUsageExtensionPoint.cs]: what resets this value?

File: src/Raven.Server/Utils/CpuUsageExtensionPoint.cs [hunk 16]
```diff
@@ -0,0 +1,263 @@
...
+                    NotifyWarning($"{msg} \nTherefore the process will restart.", e);
+
+                    InitNewProcessHandle();
```
Comment (ayende) [inline src/Raven.Server/Utils/CpuUsageExtensionPoint.cs]: what if errors are thrown here?

File: src/Raven.Server/Utils/CpuUsageExtensionPoint.cs [hunk 17]
```diff
@@ -0,0 +1,263 @@
...
+                if (_process.HasExited == false)
+                {
+                    _process.Kill();
+                }
+            }
+            catch (Exception)
+            {
+                // When the process terminating, killed or exited
+            }
+            finally
```
Comment (ayende) [inline src/Raven.Server/Utils/CpuUsageExtensionPoint.cs]: This shouldn't be in the finally

File: src/Raven.Server/Utils/CpuUsageExtensionPoint.cs [hunk 18]
```diff
@@ -0,0 +1,263 @@
...
+                    NotificationSeverity.Warning);
+                _notificationCenter.Add(alert);
+            }
+            catch
+            {
+                // nothing to do if we can't report it
+            }
+        }
+
+        private void InitNewProcessHandle()
```
Comment (ayende) [inline src/Raven.Server/Utils/CpuUsageExtensionPoint.cs]: Why does this need to be separate from the `Start`?

File: src/Raven.Server/Dashboard/ThreadsInfo.cs (hunk 10 above)
Comment (ppekrol) [inline src/Raven.Server/Dashboard/ThreadsInfo.cs]: Can you explain what is the case when this is not int/long ?

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Utils/CpuUsageExtensionPoint.cs [hunk 19]
```diff
@@ -0,0 +1,263 @@
...
+                {
+                    _didRestart = true;
+                    NotifyWarning($"{msg} \nTherefore the process will restart.", e);
+
+                    InitNewProcessHandle();
+                    Start();
+                }
+                else
+                {
+                    NotifyWarning($"{msg} \nTherefore the process will terminate.", e);
```
Comment (ppekrol) [inline src/Raven.Server/Utils/CpuUsageExtensionPoint.cs]: Need to use `Environment.NewLine` instead of `\n`

File: src/Raven.Server/Utils/CpuUsageExtensionPoint.cs (hunk 14 above)
Comment (ppekrol) [inline src/Raven.Server/Utils/CpuUsageExtensionPoint.cs]: Need to use `Environment.NewLine` instead of `\n`

File: src/Raven.Server/Utils/CpuUsageExtensionPoint.cs [hunk 20]
```diff
@@ -0,0 +1,263 @@
...
+            };
+
+            InitNewProcessHandle();
+        }
+
+        public void Start()
+        {
+            try
+            {
+                _process.Start();
```
Comment (ppekrol) [inline src/Raven.Server/Utils/CpuUsageExtensionPoint.cs]: Wouldn't it be more convenient for the user if the script would be started every X seconds and we would only read one line from it?

File: src/Raven.Server/Utils/CpuUsageExtensionPoint.cs (hunk 13 above)
Comment (ppekrol) [inline src/Raven.Server/Utils/CpuUsageExtensionPoint.cs]: Is JSON a way to go when it comes to scripts? It is not a commonly used format.

I would expect output like:
```
80.06 3.0
```

File: src/Raven.Server/RavenServer.cs [hunk 21]
```diff
@@ -296,6 +294,24 @@ void ConfigureKestrel(KestrelServerOptions options)
...
+        private void RunCpuUsageExtensionPoint()
+        {
+            if (string.IsNullOrEmpty(Configuration.Monitoring.CpuUsageMonitorExec))
+            {
+                return;
+            }
+
+            var inspector = new CpuUsageExtensionPoint(
+                _tcpContextPool,
+                Configuration.Monitoring.CpuUsageMonitorExec,
```
Comment (ppekrol) [inline src/Raven.Server/RavenServer.cs]: We probably need to escape this?

File: src/Raven.Server/RavenServer.cs [hunk 22]
```diff
@@ -296,6 +294,24 @@ void ConfigureKestrel(KestrelServerOptions options)
...
+                Configuration.Monitoring.CpuUsageMonitorExecArguments,
```
Comment (ppekrol) [inline src/Raven.Server/RavenServer.cs]: We probably need to escape this?

File: src/Raven.Server/Utils/CpuUsageExtensionPoint.cs [hunk 23]
```diff
@@ -0,0 +1,263 @@
...
+            }
+            catch (Exception e)
+            {
+                NotifyWarning("Could not start cpu usage extension point process", e);
+
+                Dispose();
```
Comment (ppekrol) [inline src/Raven.Server/Utils/CpuUsageExtensionPoint.cs]: Need to rethrow here?

**Review by ppekrol — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Dashboard/ThreadsInfo.cs (hunk 10 above)
Comment (ayende) [reply src/Raven.Server/Dashboard/ThreadsInfo.cs]: The problem is when running in k8s, when you can assign 0.5 cores

File: src/Raven.Server/Utils/CpuUsageExtensionPoint.cs (hunk 20 above)
Comment (ayende) [reply src/Raven.Server/Utils/CpuUsageExtensionPoint.cs]: No, we want this to be long lived, otherwise there are a lot of extra costs.
In particular, the common scenario we see for this is making requests to a k8s controller, which means that it will need to auth, which means that we want to have connection pooling, etc.

File: src/Raven.Server/Utils/CpuUsageExtensionPoint.cs (hunk 13 above)
Comment (ayende) [reply src/Raven.Server/Utils/CpuUsageExtensionPoint.cs]: We expect this to be a full fledged program, not just a script.
And we want to make sure that we are future proof here.

File: src/Raven.Server/Utils/CpuUsageExtensionPoint.cs (hunk 2 above)
Comment (haludi) [reply src/Raven.Server/Utils/CpuUsageExtensionPoint.cs]: By "error handling" do you mean exception from `StreamReader` method?

File: src/Raven.Server/Utils/CpuUsage.cs (hunk 3 above)
Comment (haludi) [reply src/Raven.Server/Utils/CpuUsage.cs]: @grisha-kotler Why? If I will get 0 cores I want the same behavior. To return the last  Cpu usage calculated or 0

File: src/Raven.Server/Utils/CpuUsage.cs (hunk 5 above)
Comment (haludi) [reply src/Raven.Server/Utils/CpuUsage.cs]: @maximburyak The `CpuUsageExtensionPoint` should be set once in the server initialization. Maybe if I will pass it to a function it will make more sense?

File: src/Raven.Server/Utils/CpuUsage.cs (hunk 6 above)
Comment (haludi) [reply src/Raven.Server/Utils/CpuUsage.cs]: @maximburyak But the situation in the first place is the "normal" calculation is wrong.
In my humble opinion, it is better to display 0 that show you something is wrong

File: src/Raven.Server/Utils/CpuUsage.cs (hunk 11 above)
Comment (haludi) [reply src/Raven.Server/Utils/CpuUsage.cs]: @ayende This can happen when: 
* There were issues in the extension point process and there was one try to restart
*  If the process returns a negative value.
What do you mean by abort? Just to kill the process? In the first scenario, it is already the situation. In the second scenario if I check for bottom edge maybe to check the upper side?

File: src/Raven.Server/Utils/CpuUsageExtensionPoint.cs (hunk 15 above)
Comment (haludi) [reply src/Raven.Server/Utils/CpuUsageExtensionPoint.cs]: @ayende Nothing, the implementation now is if the process has an issue the system trying to restart it once and no more

File: src/Raven.Server/Utils/CpuUsageExtensionPoint.cs (hunk 16 above)
Comment (haludi) [reply src/Raven.Server/Utils/CpuUsageExtensionPoint.cs]: @ayende The `InitNewProcessHandle` function just set the new process object properties and `Start` function handles itself

File: src/Raven.Server/Utils/CpuUsageExtensionPoint.cs (hunk 17 above)
Comment (haludi) [reply src/Raven.Server/Utils/CpuUsageExtensionPoint.cs]: @ayende I should not catch the exception?

File: src/Raven.Server/Utils/CpuUsageExtensionPoint.cs (hunk 23 above)
Comment (haludi) [reply src/Raven.Server/Utils/CpuUsageExtensionPoint.cs]: @ppekrol As I understand it the error handling should be soft so I just notify there is a problem. What do you think? how the error should be handled?

Comment (ppekrol) [reply src/Raven.Server/Utils/CpuUsageExtensionPoint.cs]: Then you probably want to return here, not 'ReadProcess' ? (check @maximburyak comment)

File: src/Raven.Server/Dashboard/ThreadsInfo.cs (hunk 10 above)
Comment (ayende) [reply src/Raven.Server/Dashboard/ThreadsInfo.cs]: This still needs to be changed. The translation from partial units to integer number of cores should be done in the script, not by RavenDB

File: src/Raven.Server/RavenServer.cs (hunk 21 above)
Comment (ayende) [reply src/Raven.Server/RavenServer.cs]: Yes, we do

File: src/Raven.Server/RavenServer.cs (hunk 22 above)
Comment (ayende) [reply src/Raven.Server/RavenServer.cs]: Yes

File: src/Raven.Server/Utils/CpuUsage.cs (hunk 11 above)
Comment (ayende) [reply src/Raven.Server/Utils/CpuUsage.cs]: Yes, on bad input from the script, kill it and retry.

File: src/Raven.Server/Utils/CpuUsage.cs (hunk 6 above)
Comment (ayende) [reply src/Raven.Server/Utils/CpuUsage.cs]: 0 is _not_ wrong.
0 is a valid value

Comment (haludi) [reply src/Raven.Server/Utils/CpuUsage.cs]: So should I return a negative value?

File: src/Raven.Server/Dashboard/ThreadsInfo.cs (hunk 10 above)
Comment (haludi) [reply src/Raven.Server/Dashboard/ThreadsInfo.cs]: In a situation where the number of cores is a fraction, the process cpu calculation will be wrong. 
To calculate the process cpu usage we compare the license limit vs the actual number of cores and use the min.

Comment (haludi) [reply src/Raven.Server/Dashboard/ThreadsInfo.cs]: https://github.com/ravendb/ravendb/pull/7960#discussion_r232563223

File: src/Raven.Server/Utils/CpuUsage.cs (hunk 11 above)
Comment (haludi) [reply src/Raven.Server/Utils/CpuUsage.cs]: https://github.com/ravendb/ravendb/pull/7960#discussion_r232561460
And to return a negative value or keep it 0?

File: src/Raven.Server/Utils/CpuUsageExtensionPoint.cs (hunk 17 above)
Comment (haludi) [reply src/Raven.Server/Utils/CpuUsageExtensionPoint.cs]: I want to restart or dispose the object while notifying the problem any way

File: src/Raven.Server/Utils/CpuUsage.cs (hunk 6 above)
Comment (ayende) [reply src/Raven.Server/Utils/CpuUsage.cs]: Yes, and make sure that this is user visible

File: src/Raven.Server/Dashboard/ThreadsInfo.cs (hunk 10 above)
Comment (ayende) [reply src/Raven.Server/Dashboard/ThreadsInfo.cs]: That is not our responsibility. That is what the script is for.
It would be doing that translation

Comment (ayende) [reply src/Raven.Server/Dashboard/ThreadsInfo.cs]: To give more context, this is important because "core" isn't well defined.
A Raspberry Pi has 4 cores, or example, but my laptop can probably beat all of them together with just one core.

The script defines what a core means in this context, not the system. 
We do ensure that we aren't going over the number of licensed core, but that is it.

File: src/Raven.Server/Utils/CpuUsageExtensionPoint.cs (hunk 23 above)
Comment (haludi) [reply src/Raven.Server/Utils/CpuUsageExtensionPoint.cs]: In this case, the origin methods to calculate CPU usage give wrong values so if the process didn't start or if it failed while running I still want to read the data from the disposed object to get error value from it. Another option I can think of is to switch the `CpuUsageExtensionPoint` to "Failed object" that will do this work and use it is I can't start. Is it make sense?

File: src/Raven.Server/Dashboard/ThreadsInfo.cs (hunk 10 above)
Comment (iftahbe) [reply src/Raven.Server/Dashboard/ThreadsInfo.cs]: We calculate machine usage and process usage separately. If you want to delegate all that responsibility to the script, it will have to give us RavenDB's process usage as well. So the script will have to monitor the RavenDB process. It makes it much harder than just calling the K8s API metrics-server.

File: src/Raven.Server/Utils/CpuUsage.cs (hunk 5 above)
Comment (maximburyak) [reply src/Raven.Server/Utils/CpuUsage.cs]: Maybe the CpuUsage class shouldn't be a "static" class, but a singleton? And this way you'll initialize all the values you need inside the constructor? It looks to me that it lacks structure and state this way..

File: src/Raven.Server/Dashboard/ThreadsInfo.cs (hunk 10 above)
Comment (ayende) [reply src/Raven.Server/Dashboard/ThreadsInfo.cs]: I pretty sure that we have to. 
And we can do that in the same way we do now, just in the script

File: src/Raven.Server/Utils/CpuUsageExtensionPoint.cs (hunk 23 above)
Comment (ppekrol) [reply src/Raven.Server/Utils/CpuUsageExtensionPoint.cs]: If process died, then you want to `ReadErrors`, not the normal output I think?

File: src/Raven.Server/Utils/CpuUsageExtensionPoint.cs [hunk 24]
```diff
@@ -0,0 +1,284 @@
...
+                    if (_process.HasExited == false)
+                    {
+                        _process.Kill();
+                    }
+                }
+                catch (Exception)
+                {
+                    // When the process terminating, killed or exited
+                }
+                errors = errors + await _process.StandardError.ReadToEndAsync();
```
Comment (ppekrol) [inline src/Raven.Server/Utils/CpuUsageExtensionPoint.cs]: Can't we read to end at the beginning?

File: src/Raven.Server/Utils/CpuUsageExtensionPoint.cs [hunk 25]
```diff
@@ -0,0 +1,284 @@
...
+            if (data == null)
+            {
+                HandleError("The output stream of the process has closed.");
+                return false;
+            }
+            using (_contextPool.AllocateOperationContext(out var context))
+            {
+                try
+                {
+                    var blittable = context.ReadForMemory(data, "cpuUsageExtensionPointData");
```
Comment (ppekrol) [inline src/Raven.Server/Utils/CpuUsageExtensionPoint.cs]: We need to dispose this

File: src/Raven.Server/Utils/CpuUsageExtensionPoint.cs (hunk 24 above)
Comment (haludi) [reply src/Raven.Server/Utils/CpuUsageExtensionPoint.cs]: As I understand it the `ReadToEndAsync` would not return until the stream is closed.
So if the process send error without exiting it will continue to wait for the continuation

File: src/Raven.Server/Utils/CpuUsageExtensionPoint.cs (hunk 23 above)
Comment (haludi) [reply src/Raven.Server/Utils/CpuUsageExtensionPoint.cs]: There is a task that responsible to read errors, notify them and kill the process if needed. But after that or if the process killed/existed by another reason when for instance the Dashboard requests for CPU usage it will get negative values to indicate the problem

**Comment by ppekrol** [comment]
Don't mind the random `System.Net.Sockets.SocketException : The I/O operation has been aborted because of either a thread exit or an application request` it is related to https://github.com/ravendb/ravendb/pull/8028

File: src/Raven.Server/Dashboard/ThreadsInfo.cs (hunk 10 above)
Comment (ayende) [inline src/Raven.Server/Dashboard/ThreadsInfo.cs]: This is still double, and should be int

File: src/Raven.Server/Utils/Cpu/CpuUsage.cs [hunk 26]
```diff
@@ -0,0 +1,185 @@
...
+using Sparrow.Json;
+using Sparrow.Logging;
+using Sparrow.Platform;
+using Sparrow.Platform.Posix;
+using Sparrow.Platform.Posix.macOS;
+using Sparrow.Utils;
+
+namespace Raven.Server.Utils.Cpu
+{
+    public static class CpuUsage
```
Comment (ayende) [inline src/Raven.Server/Utils/Cpu/CpuUsage.cs]: This should be an instance class, not a static class. 
The problem is that in the tests, we have multiple raven servers, and you are going to impact all of them in this way

File: src/Raven.Server/RavenServer.cs [hunk 27]
```diff
@@ -1654,6 +1656,7 @@ public void Dispose()
                 ea.Execute(() => ServerMaintenanceTimer?.Dispose());
                 ea.Execute(() => AfterDisposal?.Invoke());
                 ea.Execute(() => _clusterMaintenanceWorker?.Dispose());
+                ea.Execute(CpuUsage.Dispose);
```
Comment (ayende) [inline src/Raven.Server/RavenServer.cs]: Only dispose the local instance, not a static one

File: src/Raven.Server/Utils/Cpu/CpuUsage.cs [hunk 28]
```diff
@@ -0,0 +1,185 @@
...
+        [StructLayout(LayoutKind.Sequential)]
+        internal struct FileTime
+        {
+            public int dwLowDateTime;
+            public int dwHighDateTime;
+        }
+
+        [return: MarshalAs(UnmanagedType.Bool)]
+        [DllImport("kernel32.dll", SetLastError = true)]
+        internal static extern bool GetSystemTimes(
```
Comment (ayende) [inline src/Raven.Server/Utils/Cpu/CpuUsage.cs]: Why is this here and not in the Windows one?

File: src/Raven.Server/Utils/Cpu/CpuUsageCalculator.cs [hunk 29]
```diff
@@ -0,0 +1,265 @@
...
+    internal interface ICpuUsageCalculator : IDisposable
+    {
+        (double MachineCpuUsage, double ProcessCpuUsage) Calculate();
+        void Init();
+    }
+
+    [SuppressMessage("ReSharper", "StaticMemberInGenericType")]
+    internal abstract class CpuUsageCalculator<T> : ICpuUsageCalculator where T : ProcessInfo
+    {
+        protected static readonly Logger Logger = LoggingSource.Instance.GetLogger<MachineResources>("Server");
```
Comment (ayende) [inline src/Raven.Server/Utils/Cpu/CpuUsageCalculator.cs]: Why _do_ you have multiple static members here for each type?

File: src/Raven.Server/Utils/Cpu/CpuUsageCalculator.cs [hunk 30]
```diff
@@ -0,0 +1,265 @@
...
+            if (CpuUsage.GetSystemTimes(ref systemIdleTime, ref systemKernelTime, ref systemUserTime) == false)
+            {
+                if (Logger.IsInfoEnabled)
+                    Logger.Info("Failure when trying to get GetSystemTimes from Windows, error code was: " + Marshal.GetLastWin32Error());
+                return null;
+            }
+
+            return new WindowsInfo
+            {
+                SystemIdleTime = CpuUsage.GetTime(systemIdleTime),
```
Comment (ayende) [inline src/Raven.Server/Utils/Cpu/CpuUsageCalculator.cs]: `CpuUsage.GetTime` move to this class

File: src/Raven.Server/Utils/Cpu/CpuUsageExtensionPoint.cs [hunk 31]
```diff
@@ -0,0 +1,277 @@
...
+            MachineCpuUsage = -1
+        };
+
+        public ExtensionPointData Data
+        {
+            get
+            {
+                if (IsDisposed && _logger.IsOperationsEnabled)
+                {
+                    _logger.Operations($"Try to get data from {nameof(CpuUsageExtensionPoint)} which was disposed");
```
Comment (ayende) [inline src/Raven.Server/Utils/Cpu/CpuUsageExtensionPoint.cs]: This will be called all the time, which will flood the log

File: src/Raven.Server/Utils/Cpu/CpuUsageExtensionPoint.cs [hunk 32]
```diff
@@ -0,0 +1,277 @@
...
+                    if (_process.HasExited == false)
+                    {
+                        _process.Kill();
+                    }
+                }
+                catch (Exception)
+                {
+                    // When the process terminating, killed or exited
+                }
+                errors = errors + await _process.StandardError.ReadToEndAsync();
```
Comment (ayende) [inline src/Raven.Server/Utils/Cpu/CpuUsageExtensionPoint.cs]: need new line here

File: src/Raven.Server/Utils/Cpu/CpuUsageExtensionPoint.cs [hunk 33]
```diff
@@ -0,0 +1,277 @@
...
+                {
+                    using (var blittable = context.ReadForMemory(data, "cpuUsageExtensionPointData"))
+                    {
+                        if (TryGetValue(blittable, nameof(ExtensionPointData.MachineCpuUsage), out var machineCpuUsage)
+                            && TryGetValue(blittable, nameof(ExtensionPointData.ProcessCpuUsage), out var processCpuUsage))
+                        {
+                            _data.MachineCpuUsage = machineCpuUsage;
+                            _data.ProcessCpuUsage = processCpuUsage;
+                            return true;
+                        }
```
Comment (ayende) [inline src/Raven.Server/Utils/Cpu/CpuUsageExtensionPoint.cs:177]: comment here that `HandleError` is called from inside the `TryGetValue`

File: src/Raven.Server/Utils/Cpu/CpuUsageExtensionPoint.cs [hunk 34]
```diff
@@ -0,0 +1,277 @@
...
+                catch (Exception e)
+                {
+                    HandleError($"Unable to parse \"{data}\" to json", e);
+                }
+            }
+
+            return false;
+        }
+
+        private bool TryGetValue(BlittableJsonReaderObject blittable, string machineCpuUsageName, out double cpuUsage)
```
Comment (ayende) [inline src/Raven.Server/Utils/Cpu/CpuUsageExtensionPoint.cs]: `TryGetValue` is a generic name, and you are also doing error handling here, call it as such

File: src/Raven.Server/Utils/Cpu/CpuUsageExtensionPoint.cs [hunk 35]
```diff
@@ -0,0 +1,277 @@
...
+            }
+            catch (Exception)
+            {
+                // When the process terminating, killed or exited
+            }
+            finally
+            {
+                if (_didRestart == false)
+                {
+                    _didRestart = true;
```
Comment (ayende) [inline src/Raven.Server/Utils/Cpu/CpuUsageExtensionPoint.cs]: `_didRestart` change this to a `DateTime` and allow it if `DateTime.UtcNow - _lastRestart` is greater than 15 minutes

File: src/Raven.Server/Utils/ThreadsUsage.cs [hunk 36]
```diff
@@ -40,7 +41,7 @@ public ThreadsInfo Calculate()
                 var activeCores = CpuUsage.GetNumberOfActiveCores(process);
                 threadsInfo.ActiveCores = activeCores;
 
-                if (timeDiff == 0 || activeCores == 0)
+                if (timeDiff == 0 || Math.Abs(activeCores) < 0.01)
```
Comment (ayende) [inline src/Raven.Server/Utils/ThreadsUsage.cs]: activeCores should not be a double

**Review by ayende — CHANGES_REQUESTED** [review]

**Review by ppekrol — APPROVED** [review]
